### ISSUE TITLE
Feature Request: Add automatic download retry option

### ISSUE THREAD

**Issue by jerelevine** [SDWebImage/SDWebImage#2587]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 11.0 / 10.13.0 / 11.0 / 4.0
 SDWebImage Version      | e.g. 4.2.0 / 4.1.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 9 / Xcode 8
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

Feature Request:

I would love an easy way to automatically add an automatic download retry. I think that this could be an option passed to the image manager and possibly include a delegate that checks to see if the retry should occur based on the error and the retry count. I think this would be a great addition to the 5.x version of the SDK. Thanks!

**Comment by dreampiggy** [comment]
Seems a interesting feature request. I previously have a idea about it, but after some consideration I think this feature need some detailed search before writing code.

# Does the retry logic checked by framework or user
## By framework
> possibly include a delegate that checks to see if the retry should occur based on the error and the retry count

There are actuallly two choice for retry. One is that you specify enough context or option during download, then we do the automatically retry.

The API may looks like this one:

```
/**
 A NSUInteger value to specify that the total retry count when image download process failed. If the retry happended, the operation will not been marked finished and retry with the same args, until it success or the retry times reach this limit value. (NSUInteger)
 Defatuls to 0, which means does not do any retry.
 */
FOUNDATION_EXPORT SDWebImageContextOption _Nonnull const SDWebImageContextDownloadRetryCount;
```

Then you just need to pass this retry count limit value to the image request you want. It's a per-image-request config.


## By user
That simple retry count limit is one choice. Another choice it's to let the user do the check logic. It's OK to add a delegate method for manager. It's also OK to add it into downloader level since it's not related to cache, but that will cause downloader code become more complicated and does not benefit by using [Custom Loader](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#custom-loader-50) (Each loader need write code to implements it).

A API may looks like this one:

```objective-c
/**
 Control the retry download logic to mark as finished.

 @param imageManager The current `SDWebImageManager`
 @param imageURL The url of the image
 @param error The download error for the url
 @param retryTimes The current retry times for current image load, the first time is 0. And will be increased each time you do retry.
 @param operation The load operation, you can use `loaderOperation` to grab the download operation (such as `SDWebImageDownloadToken`) to check more detailed information if needed.
 @return Return true to do a retry automatically, false to mark it finally as finished
 */
- (BOOL)imageManager:(nonnull SDWebImageManager *)imageManager shouldRetryDownloadForURL:(nonnull NSURL *)imageURL withError:(nonnull NSError *)error retryTimes:(NSUInteger)retryTimes operation:(nonnull SDWebImageCombinedOperation *)operation;
```


# Does the retry logic need to change original URL ?
This is also important topic to know. Because retry for same URL multiple times may be not suitable for some logic. For example, A CDN image URL may have multiple backup host, one image URL was failed does not means all host should be failed. So it's a use case that we want to change a new URL instead.

If this feature is applied, the above API should also been updated. User must return a new URL for this. Or maybe we can tweak the API to let user combine the existing request modifier
`SDWebImageDownloaderRequestModifier` logic (When retry count == 0, return first URL host, retry count == 1, return second URL host). Anyway, this need some research before writing code.

**Comment by jerelevine** [comment]
Commenting to keep this request alive.

**Comment by jerelevine** [comment]
Not stale

**Comment by dreampiggy** [comment]
@jerelevine Any good idear ?

Actually, from myself's use case, I don't have a really strong rely on this automatically retry. Do you have some good use case for your own Apps ? Which can help me to understand and choose the proper solution (using delegation mode or the context mode, or even better) to solve this problem.

**Comment by mikemike396** [comment]
Any progress on this request? Thanks

**Comment by dreampiggy** [comment]
Any idea for this feature request ? @Insofan 

There are still 2 options here:
1. Provide the API which receive context, and policy, to manage the retry automatically. User choose the policy and rule.
2. Provide the API which expose the timing for you to retry, we as SDK don't do anything. User implements the timing and complicated rules for rety.

**Comment by dreampiggy** [comment]
Note: Kingfisher choose the first one: https://github.com/onevcat/Kingfisher/pull/1447
